logstash-core/src/main/java/org/logstash/ackedqueue/io/FileCheckpointIO.java

```diff
@@ -104,20 +107,19 @@ public void write(String fileName, Checkpoint checkpoint) throws IOException {
             out.getFD().sync();
         }
 
+        // Windows can have problem doing file move See: https://github.com/elastic/logstash/issues/12345
+        // retry a couple of times to make it works. The first two runs has no break. The rest of reties are exponential backoff.
```

Comment:
I think that the full try statement could be extracted in a `RetryStrategy`class, which could bring in input `retry` flag and internally use the  `ExponentialBackoff`.


```java
retryStrategy.apply(() -> Files.move(tmpPath, dirPath.resolve(fileName), StandardCopyOption.ATOMIC_MOVE));
.
.
.
class  RetryStrategy {
    private final ExponentialBackoff backoff;

    RetryStrategy(boolean retry) {
        this.retry = retry;
        this.backoff = new ExponentialBackoff(3L);
    }

    void apply(CheckedSupplier<T> action) {
        try {
            action.get();
        } catch (IOException ex) {
            if (retry) {
                try {
                    backoff.retry(action);
                } catch (ExponentialBackoff.RetryException re) {
                    throw new IOException("Error writing checkpoint", re);
                }
            } else {
                logger.error("Error writing checkpoint without retry: " + ex);
                throw ex;
            }
        }
    }
}
```

and then instead of have the `if (retry)` we could have 2 strategy, one with backoff to be instantiated if the `retry` is false, and the other with backoff.

This is just a rough idea.

Reply:
The main reason of having `ExponentialBackoff` is that it is reusable for other retryable code. The retry strategy of checkpoint write seems to be the only use case. I am thinking is it reusable in other parts. Maybe we can have it in the future.

Reply:
Agree